### DISCUSSION TITLE
`next/image` with both `export` and `build`

### DISCUSSION THREAD

**Discussion by haf** [vercel/next.js#29020] [Help] [answered]
Can I configure the `next/image` package to use a **custom loader** in `next.config.js` to work **both** with `export` and `build`? If so, how exactly would it be done? Can a custom image function generate static images for every configured viewport breakpoint (this would a good compromise for me)? 

Otherwise, if you leave the `next.config.js` as is; when running `next export` it crashes:

```
Error: Image Optimization using Next.js' default loader is not compatible with `next export`.
  Possible solutions:
    - Use `next start` to run a server, which includes the Image Optimization API.
    - Use any provider which supports Image Optimization (like Vercel).
    - Configure a third-party loader in `next.config.js`.
    - Use the `loader` prop for `next/image`.
  Read more: https://nextjs.org/docs/messages/export-image-api
    at /Users/h/dev/haf/cool-comet/node_modules/next/dist/export/index.js:257:23
    at async Span.traceAsyncFn (/Users/h/dev/haf/cool-comet/node_modules/next/dist/telemetry/trace/trace.js:60:20)
```

**Background**

I'm using a [Netlify NextJS plugin](https://github.com/netlify/netlify-plugin-nextjs) to deploy — and also using a product called Stackbit that both statically compiles the site and also uses the Netlify NextJS plugin to generate Netlify functions for the the next/image component.

So running  `next export` doesn't mean the site is all static, only that most of it is; but we can still have the HTML source call into an image loader deployed as a Netlify function.

**The crux**

The image Netlify-function is deployed only *after* build success, after both `build` and `export` having been run.

I don't know the URL for it at build time, but I can guess and/or set the _path_: but I can't find any docs on <quote>how to configure the `next/image` package to use a custom loader in `next.config.js`?</quote> — because if I could, I could make a deploy without `export` being run, so point the `export` custom loader to this function and then enable `next export`?

[The docs](https://nextjs.org/docs/api-reference/next/image#loader) only state a loader is this:

> A custom function used to resolve URLs. Defaults to images object in next.config.js.

And gives sample

```js
const myLoader = ({ src, width, quality }) => {
  return `https://example.com/${src}?w=${width}&q=${quality || 75}`
}
```

But I'm still not sure how to configure this in `next.config.js` to enable what I need: hybrid `build` and `export` functionality. Help needed!

#### Related

- https://github.com/vercel/next.js/discussions/20152

**Comment by wilsonneto-swe** [comment, marked as answer]
Hi @haf! Great question,

when you use "export" this means your site will be full static, such as a CRA, a simple SPA, you can't use any next.js server functions, and this includes image optimization, server-side rendering, and incremental static generation...

The simplest solutions are:
- Use the simple <Img ... /> tag
- Use a third-party CDN that allows you to optimize your images (in my full static projects I like very much to use Akamai)

![image](https://user-images.githubusercontent.com/20674439/132971194-b2457ae3-1a70-4c38-9d63-e91a35ce39be.png)
 
https://nextjs.org/docs/basic-features/image-optimization

I hope it helps you.
Best regards.
